cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -320,21 +352,34 @@ private void processMetrics() {
 
         // Skip persisting if the records is empty
         if (records.isEmpty()) {
-          return;
+          return metricsProcessIntervalMillis; // Sleep for maximum time allowed if no new metrics are received
         }
 
         // Ensure there's only one thread can persist records and messageId's when the persistingFlag is false.
         // The thread set persistingFlag to true when it starts to persist.
         if (!persistingFlag.compareAndSet(false, true)) {
           LOG.trace("Cannot persist because persistingFlag is taken by another thread.");
-          return;
+          // Calculating the length of time used to emit the metrics fetched in current run.
+          // If the number of newly fetched metrics is smaller than fetcherLimit, use the sleep time after previous run.
+          // If the number of newly fetched metrics is equal to fetcherLimit, there might be more than fetcherLimit
+          // emitted during the last sleep time. Use the difference between the time of first and last metrics
+          // fetched instead.
+          long metricsEmittingTime = newMetricsCount < fetcherLimit ?
+            previousSleepMillis : topicLatestTime.get(topicIdMetaKey) - firstMetricTime;
```

Comment:
Why use the previous sleep millis? Shouldn't the sleep time simply be determined based on the the time differences between current time and the metrics record time, as well as time spent in the processing?